Source/OxyPlot/Series/XYAxisSeries.cs

```diff
@@ -750,62 +750,65 @@ protected int UpdateWindowStartIndex<T>(IList<T> items, Func<T, double> xgetter,
         /// </returns>
-        protected int FindWindowStartIndex<T>(IList<T> items, Func<T, double> xgetter, double targetX, int initialGuess)
+        public int FindWindowStartIndex<T>(IList<T> items, Func<T, double> xgetter, double targetX, int initialGuess)
         {
-            int lastguess = 0;
             int start = 0;
-            int end = items.Count - 1;
-            int curGuess = initialGuess;
+            int nominalEnd = items.Count - 1;
+            while (nominalEnd > 0 && double.IsNaN(xgetter(items[nominalEnd])))
+                nominalEnd -= 1;
+            int end = nominalEnd;
+            int curGuess = Math.Max(0, Math.Min(end, initialGuess));
 
-            while (start <= end)
+            double GetX(int index)
             {
-                if (curGuess < start)
+                while (index <= nominalEnd)
                 {
-                    return lastguess;
-                }
-                else if (curGuess > end)
-                {
-                    return end;
+                    double guessX = xgetter(items[index]);
+                    if (double.IsNaN(guessX))
+                        index += 1;
+                    else
+                        return guessX;
                 }
+                return xgetter(items[nominalEnd]);
+            }
 
-                double guessX = xgetter(items[curGuess]);
+            while (start < end)
+            {
+                double guessX = GetX(curGuess);
                 if (guessX.Equals(targetX))
                 {
-                    return curGuess;
+                    start = curGuess;
+                    break;
                 }
                 else if (guessX > targetX)
                 {
                     end = curGuess - 1;
-                    if (end < start)
-                    {
-                        return lastguess;
-                    }
-                    else if (end == start)
-                    {
-                        return end;
-                    }
                 }
                 else
-                {
-                    start = curGuess + 1;
-                    lastguess = curGuess;
+                { 
+                    start = curGuess;
                 }
 
                 if (start >= end)
                 {
-                    return lastguess;
+                    break;
                 }
 
-                double endX = xgetter(items[end]);
-                double startX = xgetter(items[start]);
+                double endX = GetX(end);
+                double startX = GetX(start);
 
                 var m = (end - start + 1) / (endX - startX);
+                
                 curGuess = start + (int)((targetX - startX) * m);
+                curGuess = Math.Max(start + 1, Math.Min(curGuess, end));
             }
 
-            return lastguess;
+            while (start > 0 && (xgetter(items[start]) > targetX))
+                start -= 1;
```

Comment:
This condition is never fired in the tests. Is it necessary? Is it liable to obscure bugs in the main mechanism?

Reply:
Sorry, I threw this together for fixing an semi-urgent issue in my day-job. I didn't have to time to convince myself that the code would always give the correct value so added that as a sort of fail safe.

It's a horrible kludge! :-) 

It should be cleaned up. Did you run a code coverage tool? And the test didn't trigger it? I be it could be removed in that case.

Reply:
I couldn't convince myself it was needed, and I wanted to check it wasn't always being over-used (e.g. something was wrong with binary search and it was just falling back on a linear search) so I replaced the insides with a `throw` and all the tests still passed.